src/Resources/contao/controllers/BackendMain.php

```diff
@@ -293,4 +279,48 @@ protected function output()
 
 		return $this->Template->getResponse();
 	}
+
+	/**
+	 * Adjusts the logout link if the current user is impersonated.
+	 */
+	private function setImporsonatedLogout()
+	{
+		$token = \System::getContainer()->get('security.token_storage')->getToken();
+
+		if (!$token instanceof TokenInterface)
+		{
+			return;
+		}
+
+		$impersonatorUser = null;
+		foreach ($token->getRoles() as $role)
+		{
+			if ($role instanceof SwitchUserRole)
+			{
+				$impersonatorUser = $role->getSource()->getUsername();
+				break;
+			}
+		}
+
+		if (!$impersonatorUser)
+		{
+			return;
+		}
+
+		$firewallMap = \System::getContainer()->get('security.firewall.map');
+		$request = \System::getContainer()->get('request_stack')->getCurrentRequest();
+
+		// generate exit impersonation path from current request
+		if (null === $request || null === ($firewallConfig = $firewallMap->getFirewallConfig($request)) || null === ($switchUserConfig = $firewallConfig->getSwitchUser()))
+		{
+			return;
+		}
+
+		$exitPath = $request->getRequestUri();
```

Comment:
Looks a lot like you could use the `Uri` component of `League` we already use?

Reply:
Use for what?